Add Dreamdrive64 Support

khill25 opened the pull request:
This adds support for Dreamdrive64 rom cart. It's using a different address space from Picocart64, so I took a different abbreviation.

Commit:
Add picocart64/dreamdrive64 sd read support

Commit:
Added dreamdrive64 support. Renamed namespace to match different project.

Commit:
Merge branch 'trunk' into pc64/add-picocart64-support

Commit:
Removed unused function def from debugging.

Commit:
Reverted some changes that shouldn't be in the code. Used while debugging.

rasky commented:
Hi, thanks for the PR. First thing before review: is it possible to relicense the code using the Unlicense license that libdragon uses? I wouldn't want to bring in code that adds attribution constraints to our users like the BSD license does.

khill25 commented:
> Hi, thanks for the PR. First thing before review: is it possible to relicense the code using the Unlicense license that libdragon uses? I wouldn't want to bring in code that adds attribution constraints to our users like the BSD license does.

No problem. I'm used to submitting PRs internally at my previous companies 😅

Commit:
Removed license that was mistakenly placed on file.

Polprzewodnikowy commented on src/debug.c:
```diff
@@ -226,7 +227,7 @@ DWORD get_fattime(void)
 
 static DRESULT fat_disk_ioctl_default(BYTE cmd, void* buff)
-{
+{	
```
What's up with those random whitespaces added throughout the code?

Polprzewodnikowy commented on src/debug.c:
```diff
@@ -264,6 +265,16 @@ static fat_disk_t fat_disk_sc64 =
+    fat_disk_initialize_ddr64,
+    fat_disk_status_default,
+    fat_disk_read_ddr64,
+	NULL,
```
Inconsistent whitespace used, rest of the file use tab indentation and I suggest sticking with one option.

anacierdem replied:
I think we need to use a tool for this at some point, this happens a lot unfortunately. Clang format does not seem to play well with some gcc extensions though

khill25 replied:
Whitespace is driving me nuts! It looks totally fine in visual studio code and in my git client, I don't notice it's messed up until I push to the PR.

anacierdem replied:
You can choose to display them if you haven't already: https://stackoverflow.com/questions/30140595/show-whitespace-characters-in-visual-studio-code

khill25 replied:
So appears that the rest of the codebase uses spaces and I'm using tabs. I'll have to convert all my tabs to spaces

Polprzewodnikowy commented on src/debug_sdfs_ddr64.c:
```diff
@@ -0,0 +1,136 @@
+
+static uint8_t ddr64_sd_wait() {
+    uint32_t timeout = 0;
+    uint32_t isBusy __attribute__((aligned(8)));
```
You seem to have copied 64drive wait function from `usb.c`. Alignment on this variable is actually not needed as it is used only for return value from `io_read` function. USB lib has been rewritten recently and those alignment attributes were removed where possible.

Polprzewodnikowy commented on src/debug_sdfs_ddr64.c:
```diff
@@ -0,0 +1,136 @@
+static uint8_t ddr64_sd_wait() {
+    uint32_t timeout = 0;
+    uint32_t isBusy __attribute__((aligned(8)));
+	isBusy = 1;
```
Unnecessary assignment, value is overwritten by `io_read` function below.

Polprzewodnikowy commented on src/debug_sdfs_ddr64.c:
```diff
@@ -0,0 +1,136 @@
+        
+        // Took too long, abort
+        if((timeout++) > 10000000) {
+			fprintf(stdout, "SD_WAIT timed out. isBusy: %ld\n", isBusy);
```
Do we really want to have error printing in function that can return error code?

Polprzewodnikowy commented on src/debug_sdfs_ddr64.c:
```diff
@@ -0,0 +1,136 @@
+		}
+    }
+    while(isBusy != 0);
+    (void) timeout; // Needed to stop unused variable warning
```
Unnecessary, `timeout` variable is used in the code above.

Polprzewodnikowy commented on src/debug_sdfs_ddr64.c:
```diff
@@ -0,0 +1,136 @@
+	do {
+		uint64_t current_sector = sector + i;
+		
+		uint32_t part0 __attribute__((aligned(8)));
```
Again, alignment unnecessary because you're just passing value to the `io_write` function that doesn't have any alignment requirements.

Polprzewodnikowy commented on src/debug_sdfs_ddr64.c:
```diff
@@ -0,0 +1,136 @@
+		uint64_t current_sector = sector + i;
+		
+		uint32_t part0 __attribute__((aligned(8)));
+		part0 = (current_sector & 0xFFFFFFFF00000000LL) >> 32;
```
You can shift value first then mask it off like `(current_sector >> 32) & 0xFFFFFFFFUL`, I think it looks cleaner but it's personal preference.

Polprzewodnikowy commented on src/debug_sdfs_ddr64.c:
```diff
@@ -0,0 +1,136 @@
+		uint32_t part0 __attribute__((aligned(8)));
+		part0 = (current_sector & 0xFFFFFFFF00000000LL) >> 32;
+
+		uint32_t part1 __attribute__((aligned(8)));
```
Same as above.

Polprzewodnikowy commented on src/debug_sdfs_ddr64.c:
```diff
@@ -0,0 +1,136 @@
+		part0 = (current_sector & 0xFFFFFFFF00000000LL) >> 32;
+
+		uint32_t part1 __attribute__((aligned(8)));
+		part1 = (current_sector & 0x00000000FFFFFFFFLL);
```
You can ommit upper zeros, I think it looks cleaner but it's personal preference.

Polprzewodnikowy commented on src/debug_sdfs_ddr64.c:
```diff
@@ -0,0 +1,136 @@
+			i++;
+		} else {
+			retryCount--;
+			fprintf(stderr, "wait timeout\n");
```
Again, why print debug info here.

Polprzewodnikowy commented on src/debug_sdfs_ddr64.c:
```diff
@@ -0,0 +1,136 @@
+		}
+
+		if (retryCount <= 0) {
+			break;
```
No error code returned when read operation failed?

Polprzewodnikowy commented on src/debug_sdfs_ddr64.c:
```diff
@@ -0,0 +1,136 @@
+
+static DRESULT fat_disk_write_ddr64(const BYTE* buff, LBA_t sector, UINT count)
+{
+	fprintf(stderr, "Picocart64 does not currently support SD card writes.\n");
```
Change this to `assertf` instead of printing to the stderr.

Polprzewodnikowy commented on src/debug_sdfs_ddr64.c:
```diff
@@ -0,0 +1,136 @@
+    return RES_OK;
+}
+
+static DSTATUS fat_disk_initialize_ddr64(void) { return RES_OK; }
```
I would move this function before read/write functions as it's logically first to be called before doing actual R/W operations.

Polprzewodnikowy commented on src/debug_sdfs_ddr64.c:
```diff
@@ -0,0 +1,136 @@
+#define DDR64_MAGIC                  (0xD9D96400)
+
+// [WRITE]: Write number of bytes to print from TX buffer
+#define DDR64_REGISTER_UART_TX       (0x00000004)
```
Don't really understand why unused defines are added here.

khill25 commented:
Wow thanks for the though review! It's been a while and I let my code standards lapse 😅 I'll set about getting this fixed up. Not sure what is up with the whitespace, I'll run a formatter and see if I can sort that out.

Commit:
Fixed whitespace using visual studio code's trim whitespace feature (hopefully that works).
Removed debug print statements from Dreamdrive64 files.
Cleaned up formatting.

Commit:
Fix - whitespace looks fine in visual studio code but pushed to github it doesn't.

Commit:
More whitespace fixes

Commit:
whitespace

Commit:
Convert tabs to spaces, I'm so sorry.

khill25 commented:
Weep for the tab to space changes.

Anyway I think I addressed all the feedback. Again, I'm really sorry about the whitespace, I know it's annoying to review code that's basically all whitespace changes.

rasky commented:
Hi @khill25, let me update you on this PR.

Your code was sent as two major changes were being worked on. One (just landed) is #366 that is a big improvement to usb.c (coming from upstream). Another one, waiting for last touches, is #371 that basically removes the current SD code in favor of vendoring [libcart](https://github.com/devwizard64/libcart), which is a newly released library which is more comprehensive that our builtin SD support. Our goal is not having to maintain the SD code anymore in here, and offload that to libcart.

Your PR touches the same areas at the same time, so it is unfortunately due to some rework once also #371 lands. My suggestion would be to wait until #371 lands and then I can point you to the better way of reworking your code with the new overall architecture. That will probably be adding support for Dreamdrive64 to libcart rather than libdragon proper, but let me see where we end with #371 first.

Thanks for the patience!

rasky commented:
Hi @khill25, the libcart PR (#371) was merged. This means that SD flashcart code is now maintained at upstream [libcart](https://github.com/devwizard64/libcart) which is where you should send your PR now. I will close this PR now as we won't be able to merge it as-is, as we don't carry SD flashcard code anymore.
